subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy

```diff
@@ -0,0 +1,330 @@
+    }
+
+    def "background job executor runs jobs submitted at execution time"() {
+        given:
+        buildFile << """
+            import ${CompletableFuture.name}
+
+            def serviceRef = gradle.extensions.serviceRef
+            task check {
+                doLast {
+                    def service = serviceRef.get()
+                    def requiredServices = service._requiredServices
+                    CompletableFuture.runAsync({
+                        println "backgroundJobExecutor.executed"
+                    }, requiredServices.backgroundJobExecutor).exceptionally { e ->
+                        println "backgroundJobExecutor.status = \${e.cause.message}"
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        plugin.assertBackgroundJobCompletedBeforeShutdown(output, "backgroundJobExecutor.executed")
+
+        when:
+        succeeds("check")
+
+        then:
+        plugin.assertBackgroundJobCompletedBeforeShutdown(output, "backgroundJobExecutor.executed")
+    }
+
+    def "background job failure at configuration time fails build"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            requiredServices.backgroundJobExecutor.execute {
+                 throw new RuntimeException("Background job failed!")
+            }
+
+            task check {}
+        """
+
+        when:
+        fails("check")
+
+        then:
+        failure.assertHasDescription("Background job failed!")
+    }
+
+    def "background job failure at execution time fails build"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+
+            task check {
+                doLast {
+                    requiredServices.backgroundJobExecutor.execute {
+                         throw new RuntimeException("Background job failed!")
+                    }
+                }
+            }
+        """
+
+        when:
+        fails("check")
+
+        then:
+        failure.assertHasDescription("Background job failed!")
+
+        when:
+        fails("check")
+
+        then:
+        failure.assertHasDescription("Background job failed!")
+    }
+
+    def "background jobs can run in parallel with tasks on #numWorkers workers"(int numWorkers) {
+        given:
+        buildFile << """
+            import ${CountDownLatch.name}
+            import ${TimeUnit.name}
+
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+
+            def latch = new CountDownLatch(1)
+            def latch2 = new CountDownLatch(1)
```

Comment:
We'd usually use `BlockingHttpServer` instead of latches and timeouts to test parallel execution from an int test.

Reply:
Done, thanks for pointing to this class